services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/security/FileScanService.kt

```diff
@@ -0,0 +1,98 @@
+import org.springframework.web.reactive.function.BodyInserters.fromMultipartData
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.WebClientResponseException
+
+private val logger = KotlinLogging.logger {}
+
+private const val FORM_KEY = "FILES"
+
+@Service
+class FileScanService(
+    webClientBuilder: WebClient.Builder,
+    @Value("\${haitaton.clamav.baseUrl}") clamAvUrl: String,
+) {
+
+    private val fileScanClient: WebClient =
+        webClientBuilder.baseUrl(clamAvUrl).build().also {
+            logger.info { "Initialized file scan client with base-url: $clamAvUrl" }
+        }
+
+    fun scanFiles(files: List<Pair<String, ByteArray>>): List<FileResult> {
+        logger.info { "Scanning ${files.size} files." }
+
+        val data =
+            MultipartBodyBuilder()
+                .apply {
+                    files.forEach { (name, bytes) ->
+                        part(FORM_KEY, ByteArrayResource(bytes), APPLICATION_OCTET_STREAM)
+                            .filename(name)
+                    }
+                }
+                .build()
+
+        val response =
+            fileScanClient
+                .post()
+                .uri("/api/v1/scan")
+                .contentType(MULTIPART_FORM_DATA)
+                .accept(APPLICATION_JSON)
+                .body(fromMultipartData(data))
+                .retrieve()
+                .bodyToMono(FileScanResponse::class.java)
+                .timeout(ofSeconds(60))
+                .doOnError(WebClientResponseException::class.java) {
+                    logger.error { "Error uploading file: $it" }
+                }
+                .blockOptional()
+                .orElseThrow()
+
+        return extractResults(response).also { results -> logStatus(results) }
+    }
+
+    private fun extractResults(response: FileScanResponse): List<FileResult> {
+        if (!response.success) {
+            throw FileScanException("Scan failed, result: ${this.toJsonString()}")
+        }
+        return response.data.result
+    }
+
+    private fun logStatus(results: List<FileResult>) {
+        if (results.hasInfected()) {
+            results.filterInfected().forEach {
+                logger.warn { "Infected file detected, scan result: ${it.toJsonString()}" }
+            }
+        } else {
+            logger.info { "Files scanned successfully." }
+        }
+    }
+}
+
+data class FileScanResponse(val success: Boolean, val data: FileScanData)
+
+data class FileScanData(val result: List<FileResult>)
+
+data class FileResult(
+    val name: String,
+    @JsonProperty("is_infected") val isInfected: Boolean?,
+    val viruses: List<String>
+)
+
+class FileScanException(message: String) : RuntimeException(message)
+
+fun List<FileResult>.hasInfected(): Boolean = any { isNotFalse(it.isInfected) }
+
+fun List<FileResult>.filterInfected(): List<FileResult> = filter { isNotFalse(it.isInfected) }
```

Comment:
You can use simple equality here:
```suggestion
fun List<FileResult>.hasInfected(): Boolean = any { it.isInfected != false }

fun List<FileResult>.filterInfected(): List<FileResult> = filter { it.isInfected != false }
```

Reply:
Done.